src/execution/compiler/operator/hash_join_translator.cpp

```diff
@@ -252,6 +273,48 @@ void HashJoinTranslator::CheckRightMark(WorkContext *ctx, FunctionBuilder *funct
   check_condition.EndIf();
 }
 
+void HashJoinTranslator::CollectUnmatchedLeftRows(FunctionBuilder *function) const {
+  auto *codegen = GetCodeGen();
+
+  // var naiveIterBase: HashTableNaiveIterator
+  ast::Identifier naive_iter_base = codegen->MakeFreshIdentifier("naiveIterBase");
+  ast::Expr *naive_iter_type = codegen->BuiltinType(ast::BuiltinType::HashTableNaiveIterator);
+  function->Append(codegen->DeclareVarNoInit(naive_iter_base, naive_iter_type));
+
+  // var naiveIter = &naiveIterBase
+  ast::Identifier naive_iter = codegen->MakeFreshIdentifier("naiveIter");
+  ast::Expr *naive_iter_init = codegen->AddressOf(codegen->MakeExpr(naive_iter_base));
+  function->Append(codegen->DeclareVarWithInit(naive_iter, naive_iter_init));
+
+  // while (hasNext()):
+  ast::Expr *join_ht = global_join_ht_.GetPtr(codegen);
+  ast::Expr *naive_iter_expr = codegen->MakeExpr(naive_iter);
+  Loop loop(function, codegen->MakeStmt(codegen->HTNaiveIteratorInit(naive_iter_expr, join_ht)),
+            codegen->HTNaiveIteratorHasNext(naive_iter_expr),
+            codegen->MakeStmt(codegen->HTNaiveIteratorNext(naive_iter_expr)));
+  {
+    // var buildRow = @htNaiveIterGetRow()
+    function->Append(
+        codegen->DeclareVarWithInit(build_row_var_, codegen->HTNaiveIteratorGetRow(naive_iter_expr, build_row_type_)));
+
+    auto left_mark = codegen->AccessStructMember(codegen->MakeExpr(build_row_var_), build_mark_);
+
+    // If mark is true, then row was not matched
+    If check_condition(function, left_mark);
+    {
+      // // outerJoinConsumer(queryState, buildRow);
+      std::initializer_list<ast::Expr *> args{GetQueryStatePtr(),
+                                              codegen->MakeExpr(GetPipeline()->GetPipelineStateVar()),
+                                              codegen->MakeExpr(build_row_var_)};
+      function->Append(codegen->Call(outer_join_consumer_, args));
```

Comment:
Could it not be possible that while you created the pipeline for outer join consumer you would also want to make sure you called the initializepipeline and the finishpipeline for the outer join consumer's pipeline?

Reply:
I think this is taken care of. For the examples that I'm running the last operator in the pipeline is the `outputTranslator` and inside the `outputTranslator` finishpipeline is where the result is finalized and I can see the corresponding tpl (`@resultBufferFinalize(queryState.execCtx)`) being generated.